Added sigma epsilon version of Lennard-Jones potential

sansona opened the pull request:
Hi! Just added a simple function for a sigma-epsilon version of the Lennard Jones potential. Please let me know of any feedback!

arm61 commented on pylj/forcefields.py:
```diff
@@ -29,11 +29,48 @@ def lennard_jones(dr, constants, force=False):
-        return 12 * constants[0] * np.power(dr, -13) - (
-            6 * constants[1] * np.power(dr, -7)
-        )
+        return 12 * constants[0] * np.power(dr, -13) -
```
From a quick look at the travis build attempt it looks like there is an issue with the new line being immediately following the minus symbol.

bjmorgan replied:
Multiline statements need a backslash at the end of incomplete lines, or to be wrapped in brackets (preferred style): https://stackoverflow.com/questions/53162/how-can-i-do-a-line-break-line-continuation-in-python

arm61 replied:
On style, I am not fussed as the next time I push something to it I will be [black](https://github.com/python/black) formatting the whole thing. 
I will however add a comment about black formatting to the [CONTRIBUTING.md](https://github.com/arm61/pylj/blob/master/CONTRIBUTING.md) (see issue https://github.com/arm61/pylj/issues/44)

bjmorgan replied:
I was referring to usual Python convention, not `pylj` style.

sansona commented:
Thank you for your quick response and patience - it's my first pull request
& first time contributing to open source, so it's greatly appreciated. I've
fixed the syntactical errors and added some tests to confirm that all
functions are working as expected

On Wed, Apr 24, 2019 at 11:19 PM Andrew McCluskey <notifications@github.com>
wrote:

> *@arm61* requested changes on this pull request.
>
> Thank you for the pull request! This is functionality that I have desired
> for a while so I greatly appreciate it.
>
> It seems that the travis CI is failing (cause the tests aren’t passing). I
> have pinpointed the first breakpoint however before future commits it would
> be efficient to build, install and run the tests locally (see the README.md
> for how to do this).
> ------------------------------
>
> In pylj/forcefields.py
> <https://github.com/arm61/pylj/pull/42#discussion_r278407318>:
>
> > @@ -29,11 +29,48 @@ def lennard_jones(dr, constants, force=False):
>
>          The potential energy or force between the particles.
>
>      """
>
>      if force:
>
> -        return 12 * constants[0] * np.power(dr, -13) - (
>
> -            6 * constants[1] * np.power(dr, -7)
>
> -        )
>
> +        return 12 * constants[0] * np.power(dr, -13) -
>
>
> From a quick look at the travis build attempt it looks like there is an
> issue with the new line being immediately following the minus symbol.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arm61/pylj/pull/42#pullrequestreview-230464463>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEHPG22W6223CX3UTUSKGLTPSFENRANCNFSM4HII6ARQ>
> .
>


-- 

*Jiaming Chen*

arm61 commented:
Sorry for the slight delay in getting back to you. The changes look great to me and I am happy to merge them. But before I do can I ask you to add yourself to the creators in the [.zenodo.json](https://github.com/arm61/pylj/blob/master/.zenodo.json) file so that we correctly recognise your contribution!

sansona commented:
No worries, that's awesome! I added my name in. I also pushed a commit
where I added a square-well model not realizing it'd be added to the pull
request - I suspect I may have implemented the force component incorrectly
but could use your knowledge about that

On Wed, May 1, 2019 at 7:40 AM Andrew McCluskey <notifications@github.com>
wrote:

> Sorry for the slight delay in getting back to you. The changes look great
> to me and I am happy to merge them. But before I do can I ask you to add
> yourself to the creators in the .zenodo.json
> <https://github.com/arm61/pylj/blob/master/.zenodo.json> file so that we
> correctly recognise your contribution!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arm61/pylj/pull/42#issuecomment-488300752>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEHPG272MBAZ55PYN6GAPKTPTGTVBANCNFSM4HII6ARQ>
> .
>


-- 

*Jiaming Chen*

arm61 commented:
Awesome. Let me think about the square well potential and I will get back to you at the weekend.

arm61 commented:
@bjmorgan I was wondering if you could comment of the force calculation for a square well potential model?

bjmorgan commented:
@arm61 Technically the forces for a square-well potential are zero on either side of the step, and infinite at the step. If you are only interested in energies (e.g. Monte Carlo simulations) this is not an issue, but for molecular dynamics it becomes a problem. You can approximate an square-well potential with a smooth function that is continuously differentiable (see e.g. https://link.aps.org/doi/10.1103/PhysRevB.89.054304) but your energy conservation will get more fragile as you get closer to a true step, requiring shorter MD timesteps.

bjmorgan commented:
@sansona I have a  question about your square-well potential code. You have defined `dr` as an array of distances, but your code appears to expect `dr` to be a float, e.g. 
```python
if dr < constants[1]
```
I think your code as it stands will give an error if you pass in an array.

arm61 commented:
@sansona once you deal with the issue that @bjmorgan has pointed out I reckon there is two choices, either you try and and approximate the square well with some continuous function or you make it such that the square well model can only be used for Monte Carlo simulation (e.g. it throws some exception when the `force` variable is `True`

bjmorgan commented:
@arm61 Looking at the `forcefields.py` source, the existing functions appear to be able to handle numpy arrays of distances. Can you confirm that this is the expected behaviour? If so, it seems better if any new forcefields behave in the same way to allow them to interface with the main simulation code.

arm61 commented:
@bjmorgan the forcefield files should be able to handle numpy arrays (it is necessary for the simulation as the force calculation is vectorised).

@sansona my suggestion for dealing with this would be to make sure that your tests pass arrays as inputs (I might be being a bit rich by saying this as I am not sure my own tests do this, but @bjmorgan makes a good point).

bjmorgan commented:
> @sansona my suggestion for dealing with this would be to make sure that your tests pass arrays as inputs (I might be being a bit rich by saying this as I am not sure my own tests do this, but @bjmorgan makes a good point).

Can confirm they do not. 😱

arm61 commented:
> > @sansona my suggestion for dealing with this would be to make sure that your tests pass arrays as inputs (I might be being a bit rich by saying this as I am not sure my own tests do this, but @bjmorgan makes a good point).
> 
> Can confirm they do not. scream

Well that's some work for the train later for me!

bjmorgan commented:
@arm61 If you are going to update your tests, I would also put in examples where your parameters are different. If they all have the same value, you can't test for accidentally swapping e.g. sigma and epsilon inside the function.

arm61 commented:
> @arm61 If you are going to update your tests, I would also put in examples where your parameters are different. If they all have the same value, you can't test for accidentally swapping e.g. sigma and epsilon inside the function.

https://github.com/arm61/pylj/issues/43

bjmorgan commented:
@sansona I would also suggest changing:
```python
        if dr < constants[1]:
             return None
```
to 
```python
        if dr < constants[1]:
             raise ValueError('The energy is infinite for dr < sigma')
```
Returning `None` is unclear about what should happen next, the simulation could crash, or this could get converted to an energy of `0.0` and everything continues to run.

sansona commented:
Wow, thanks for all the feedback! I think my priorities will be

1. Making sure the square-well implementation is correct. I don't have
access to the paper @bjmorgan linked but would greatly appreciate a .pdf if
one's laying around! If not, I may just leave the force component throw an
exception until later date when I or someone else is up to implementing an
approximation
2. Making sure the various force functions can handle numpy arrays as
inputs & writing the appropriate tests. This'll take a back seat until #1's
complete


On Fri, May 3, 2019 at 2:37 AM Benjamin Morgan <notifications@github.com>
wrote:

> *@bjmorgan* commented on this pull request.
> ------------------------------
>
> In pylj/forcefields.py
> <https://github.com/arm61/pylj/pull/42#discussion_r280715041>:
>
> > @@ -29,11 +29,48 @@ def lennard_jones(dr, constants, force=False):
>          The potential energy or force between the particles.
>      """
>      if force:
> -        return 12 * constants[0] * np.power(dr, -13) - (
> -            6 * constants[1] * np.power(dr, -7)
> -        )
> +        return 12 * constants[0] * np.power(dr, -13) -
>
> I was referring to usual Python convention, not pylj style.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arm61/pylj/pull/42#discussion_r280715041>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEHPG24FYJFKULUZ5RRABE3PTQBTXANCNFSM4HII6ARQ>
> .
>


-- 

*Jiaming Chen*

sansona commented:
I updated the square well potential function to handle arrays as inputs and raise a ValueError if force=True

I could use your inputs on the following:
1. Do we want to raise a similar ValueError when an infinite energy is returned? I'm not too sure how pylj would handle infinite energies.
2. Just to be sure: when a single value is the input to one of the forcefield functions, the output should be similarly a single value whereas when an array is input, the output should be a numpy array?
3. Any comments/suggestions on the implementation of the square well function?

Thanks as before!

arm61 commented on pylj/forcefields.py:
```diff
@@ -140,20 +138,28 @@ def square_well(dr, constants, force=False):
-        elif dr >= constants[2] * constants[1]:
-            return 0
+        E = []
+        for r in dr:
```
I suspect that this may be achieved with np intrinsic function something like 

```
e[np.where(dr < constants[1])] = np.inf
```
etc. 

(I am not 100% on that code I am doing it from memory in a pub in sweden but I think that is the jist)

arm61 commented on pylj/forcefields.py:
```diff
@@ -140,20 +138,28 @@ def square_well(dr, constants, force=False):
+        E = []
+        for r in dr:
+            if r < constants[1]:
+                E.append(float('inf'))
```
just append `np.inf`

arm61 commented:
@sansona I like where these changes are going and I hope that @bjmorgan and I getting distracted making fun of my lazy testing isn’t distracting. 

Thinking more about the square well, I think that a finite (e.g. not infinite just some user defined number) square well would be the more pragmatic approach as I think the use of an inf might break everything...

If you drop me an email (andrew.mccluskey’at’diamond.ac.uk) I will send you @bjmorgan’s paper.

sansona commented:
Thanks! It's absolutely not a distraction - if anything, it makes working
on the project more fun now that I know you guys are invested in it and
care about the code quality :).

I went ahead and re-implemented a more efficient version of the previous
implementation. I think it's more maintainable and probably faster given
it's vectorized, but feedback and advice regarding any changes always
welcome. I also added a way for the user to set an upper bound in the
function rather than just having it default to inf and potentially break
things

On Sun, May 5, 2019 at 7:18 AM Andrew McCluskey <notifications@github.com>
wrote:

> @sansona <https://github.com/sansona> I like where these changes are
> going and I hope that @bjmorgan <https://github.com/bjmorgan> and I
> getting distracted making fun of my lazy testing isn’t distracting.
>
> Thinking more about the square well, I think that a finite (e.g. not
> infinite just some user defined number) square well would be the more
> pragmatic approach as I think the use of an inf might break everything...
>
> If you drop me an email (andrew.mccluskey’at’diamond.ac.uk) I will send
> you @bjmorgan <https://github.com/bjmorgan>’s paper.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arm61/pylj/pull/42#issuecomment-489430737>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEHPG25D5HDBPGHBR25KY5LPT3UCHANCNFSM4HII6ARQ>
> .
>


-- 

*Jiaming Chen*

arm61 commented:
I like the look of this! I am happy to pull this in and if you want to implement the force stuff at a later date that works.
